### DISCUSSION TITLE
Sharing context across backend using React.cache in a consistent, reliable way.

### DISCUSSION THREAD

**Discussion by ajwootto** [vercel/next.js#60124] [App Router]
I'm working on a library that provides a feature flagging client to app router in both Server and Client components.
The problem I'm facing is trying to come up with a way to share the instance of the feature flagging client across all server components, while allowing it to be initialized in a single location (eg. a layout)

What I want:

In the root layout:
```jsx
// layout.tsx 
export default async function Layout({children}) {
  return <html>
    <body>
      <FeatureFlaggingProvider
            sdkKey={process.env.NEXT_PUBLIC_FF_SDK_KEY ?? ''}
            user={{ user_id: '123' }}
            options={{}}
        >
            {children}
        </FeatureFlaggingProvider>
    </body>
  </html>
}
```

In a server component within the current page:
```jsx
export const ServerComponent = async () => {
    const flagVal = await getFlagValue('enabled-feature', false)

    return (
        <div>
            <h1>Server Component</h1>
            <p>Server Flag Value: {JSON.stringify(flagVal)}</p>
        </div>
    )
}
```

This gives the library similar ergonomics to familiar clientside React libraries, where you set up a "Provider" at the root of the application in one place, and then you can access its provided value anywhere in the tree.

To implement this I tried using the `cache` method from React, which allows me to store an instance of the client which is cached for the duration of the request. That fit the usecase of my library perfectly because each request has a different user and a different context for what the flag values should be, so it should get reinitialized from the configuration on each request. 

 I'm doing that in a manner similar to the server-only-context library:
https://github.com/manvalls/server-only-context

The provider component in the above example sets up the client instance in cache whenever it renders.

That appeared to work until I discovered a bunch of things that made it not work.

Since the storing of the client instance in cache is a side-effect of the Layout component rendering:
1. layout is not rendered first even though it's at the top of the tree, it appears Next renders page first
  - I worked around this by having the `getFlagValue` function check every tick for the client to appear in cache, and then proceed. This works but seems to add at least several milliseconds of latency to the request and probably breaks some kind of optimization Next is doing that justifies the weird rendering order
2. layout is not rendered on client navigation. Since Next tries to cache and re-use the layout, it will not re-execute it on a server-rendering request that is caused by clientside navigation. That means that all the `getFlagValue` calls in the rendering of the new page will not have the context they need to work.

Is there any way to achieve this "init once share everywhere" context pattern on the app router serverside? I don't want to require that the flagging client be initialized in every `page.tsx` file since it's meant to have a shared set of options that stay consistent, and the ergonomics of putting it everywhere would be really annoying. 

I thought I could use Middleware, but it seems like that is run in a different process / environment and can't share anything with the server rendering process.

Basically it doesn't seem like there is any place where you can have a piece of code consistently get called on every request to the server and set something that can be shared across the rendering tree. Is there something I'm missing?

**Comment by Luk-z** [comment]
try this approach https://dev.to/jdgamble555/easy-context-in-react-server-components-rsc-1mdf

You don't need to wrap the app in a provider.
Just call `useServerProvider('key').value="foo"` to set value and `const foo = useServerProvider('key').value` to read in your server component (layout and page should be a server component)
